samples/nrf9160/modem_shell/src/shell.c

```diff
@@ -0,0 +1,284 @@
+{
+	shell_global = shell_backend_uart_get_ptr();
+
+	switch (event->type) {
+	case LWM2M_CARRIER_EVENT_BSDLIB_INIT:
+		shell_print(shell_global, "LwM2M carrier event: modem lib initialized");
+		break;
+	case LWM2M_CARRIER_EVENT_CONNECTING:
+		shell_print(shell_global, "LwM2M carrier event: connecting");
+		k_sem_give(&nrf_modem_lib_initialized);
+		break;
+	case LWM2M_CARRIER_EVENT_CONNECTED:
+		shell_print(shell_global, "LwM2M carrier event: connected");
+		break;
+	case LWM2M_CARRIER_EVENT_DISCONNECTING:
+		shell_print(shell_global, "LwM2M carrier event: disconnecting");
+		break;
+	case LWM2M_CARRIER_EVENT_DISCONNECTED:
+		shell_print(shell_global, "LwM2M carrier event: disconnected");
+		break;
+	case LWM2M_CARRIER_EVENT_BOOTSTRAPPED:
+		shell_print(shell_global, "LwM2M carrier event: bootstrapped");
+		break;
+	case LWM2M_CARRIER_EVENT_LTE_READY:
+		shell_print(shell_global, "LwM2M carrier event: LTE ready");
+		break;
+	case LWM2M_CARRIER_EVENT_REGISTERED:
+		shell_print(shell_global, "LwM2M carrier event: registered");
+		break;
+	case LWM2M_CARRIER_EVENT_DEFERRED:
+		shell_print(shell_global, "LwM2M carrier event: deferred");
+		lwm2m_print_deferred(event);
+		break;
+	case LWM2M_CARRIER_EVENT_FOTA_START:
+		shell_print(shell_global, "LwM2M carrier event: fota start");
+		break;
+	case LWM2M_CARRIER_EVENT_REBOOT:
+		shell_print(shell_global, "LwM2M carrier event: reboot");
+		break;
+	case LWM2M_CARRIER_EVENT_ERROR:
+		shell_print(shell_global, "LwM2M carrier event: error");
+		lwm2m_print_err(event);
+		break;
+	}
+
+	return 0;
+}
+#endif
+
+#if defined(CONFIG_MOSH_IPERF3)
+static int cmd_iperf3(const struct shell *shell, size_t argc, char **argv)
+{
+	(void)iperf_main(argc, argv);
+	return 0;
+}
+#endif
+
+#if defined(CONFIG_MOSH_CURL)
+static int cmd_curl(const struct shell *shell, size_t argc, char **argv)
+{
+	(void)curl_tool_main(argc, argv);
+	shell_print(shell, "\nDONE");
+	return 0;
+}
+SHELL_CMD_REGISTER(curl, NULL, "For curl usage, just type \"curl --manual\"", cmd_curl);
+#endif
+
+SHELL_CMD_REGISTER(at, NULL, "Execute an AT command.", at_shell);
+
+#if defined(CONFIG_MOSH_SOCK)
+SHELL_CMD_REGISTER(sock, NULL,
+	"Commands for socket operations such as connect and send.",
+	sock_shell);
+#endif
+
+#if defined(CONFIG_MOSH_PING)
+SHELL_CMD_REGISTER(ping, NULL, "For ping usage, just type \"ping\"", icmp_ping_shell);
+#endif
+
+#if defined(CONFIG_MOSH_LTELC)
+SHELL_CMD_REGISTER(ltelc, NULL,
+	"Commands for LTE link controlling and status information.",
+	ltelc_shell);
+#endif
+
+#if defined(CONFIG_MOSH_IPERF3)
+SHELL_CMD_REGISTER(iperf3, NULL, "For iperf3 usage, just type \"iperf3 --manual\"", cmd_iperf3);
+#endif
+
+#if defined(CONFIG_MOSH_SMS)
+SHELL_CMD_REGISTER(sms, NULL, "Commands for sending and receiving SMS.", sms_shell);
+#endif
+
+#if defined(CONFIG_MOSH_PPP)
+SHELL_CMD_REGISTER(ppp, NULL,
+	"Commands for controlling PPP.",
+	ppp_shell_cmd);
+#endif
+
+static void disable_uarts(void)
```

Comment:
It is possible with the Device Power Management API, i.e.:
```
pm_device_state_set(uart_dev, PM_DEVICE_STATE_LOW_POWER, NULL, NULL);
```
and then `PM_DEVICE_STATE_ACTIVE` to enable back the UART.
And this will be much better than accessing the peripheral registers directly.

Reply:
@anangl Great, thanks! I'll update the code.

Reply:
Found it after PR rebase, I'll give it a try.

Reply:
@rlubos The only way to get it working with the Zephyr PM API was to enable UART1 in the device tree. Otherwise idle current consumption was higher than it should be. I think we need UART1 anyway for modem tracing and PPP, so I added a board overlay for the DK which enables UART1. Please check if this looks ok now.

Reply:
Looks good to me